tools/reviewer/job/impl/SubmitterTask.java

```diff
@@ -0,0 +1,187 @@
+
+                  boolean shouldPush = true;
+                  List<GitRepo> gitRepos = new ArrayList<>();
+
+                  CIRequest.Builder newRequestBuilder =
+                      CIRequest.newBuilder().setForSubmission(true);
+
+                  for (CIResponse.TargetResult targetResult : response.getResultsList()) {
+                    CIRequest.Target target = targetResult.getTarget();
+
+                    Repo repo = target.getRepo();
+                    String repoPath =
+                        fileUtils.joinPaths(
+                            fileUtils.getCurrentWorkingDirectory(), "submitter", repo.getId());
+                    GitRepo gitRepo = this.gitRepoFactory.create(repoPath);
+                    gitRepos.add(gitRepo);
+
+                    try {
+                      if (fileUtils.folderEmptyOrNotExists(repoPath)) {
+                        gitRepo.cloneRepo(repo.getUrl(), repoPath);
+                      } else {
+                        gitRepo.pull();
+                      }
+                    } catch (IOException e) {
+                      e.printStackTrace();
+                    }
+
+                    String branch = String.format("origin/D%d", diff.getId());
+                    gitRepo.switchBranch(branch);
+
+                    newRequestBuilder =
+                        newRequestBuilder.addTarget(
+                            CIRequest.Target.newBuilder()
+                                .setRepo(repo)
+                                .setCommitId(gitRepo.getHeadCommitId())
+                                .build());
+
+                    if (gitRepo.getHeadCommitId().equals(target.getCommitId())) {
+                      // latest result for CI matches last commit on the branch
+
+                      gitRepo.switchBranch("master");
+                      boolean mergingResult = gitRepo.merge("D%d", true);
+                      if (!mergingResult || gitRepo.hasChanges("master")) {
+                        shouldPush = false;
+                      }
+
+                    } else {
+                      // commit on the branch is newer than results in CI
+                      // it means we'll need to later push a new CIRequest
+                      shouldPush = false;
+                    }
+                  }
+
+                  if (shouldPush) {
+                    boolean allPushesSuccessful =
+                        gitRepos.stream().allMatch(repo -> repo.push("master"));
+                    if (allPushesSuccessful) {
+                      diff = diff.toBuilder().setStatus(Diff.Status.SUBMITTED).build();
+                      this.firestoreClient.createProtoDocument(
+                          DIFF_PATH, String.valueOf(diff.getId()), diff);
+                    } else {
+                      throw new RuntimeException("Not all pushes were successful");
+                    }
+                  } else {
+                    if (!ciSubmittedDiffIds.contains(diff.getId())) {
```

Comment:
Why don't we hash the `newRequestBuilder.build()` object?
We might want to run a different request for the same diff.

Reply:
@oferb 
>We might want to run a different request for the same diff.

why is that? I think `SubmitterTask` only submits new requests in this one specific case. Generally, we rely on `aa` doing it.

Reply:
@oferb makes sense!